server/src/core/xml_arch_builder_rng_validation.rs

```diff
@@ -0,0 +1,684 @@
+                                Some(EXTENSION_NAME.to_string()), 
+                                format!("submenuitems are not allowed when Action attribute and parent are specified"), 
+                                None, 
+                                None));
+                        }
+                    }
+                }
+                "parent" => {
+                    if is_submenu {
+                        diagnostics.push(Diagnostic::new(
+                            Range { start: Position::new(attr.range().start as u32, 0), end: Position::new(attr.range().end as u32, 0) }, 
+                            Some(DiagnosticSeverity::ERROR), 
+                            Some(lsp_types::NumberOrString::String(S!("OLS30408"))),
+                            Some(EXTENSION_NAME.to_string()), 
+                            format!("parent attribute is not allowed in submenuitems"), 
+                            None, 
+                            None));
+                    }
+                }
+                "web_icon" => {
+                    if has_parent || is_submenu {
+                        diagnostics.push(Diagnostic::new(
+                            Range { start: Position::new(attr.range().start as u32, 0), end: Position::new(attr.range().end as u32, 0) }, 
+                            Some(DiagnosticSeverity::ERROR), 
+                            Some(lsp_types::NumberOrString::String(S!("OLS30406"))),
+                            Some(EXTENSION_NAME.to_string()), 
+                            format!("web_icon attribute is not allowed when parent is specified"), 
+                            None, 
+                            None));
+                    }
+                }
+                _ => {
+                    diagnostics.push(Diagnostic::new(
+                        Range { start: Position::new(attr.range().start as u32, 0), end: Position::new(attr.range().end as u32, 0) }, 
+                        Some(DiagnosticSeverity::ERROR), 
+                        Some(lsp_types::NumberOrString::String(S!("OLS30403"))),
+                        Some(EXTENSION_NAME.to_string()), 
+                        format!("Invalid attribute {} in menuitem node", attr.name()), 
+                        None, 
+                        None));
+                }
+            }
+        }
+        if !found_id {
+            diagnostics.push(Diagnostic::new(
+                Range { start: Position::new(node.range().start as u32, 0), end: Position::new(node.range().end as u32, 0) }, 
+                Some(DiagnosticSeverity::ERROR), 
+                Some(lsp_types::NumberOrString::String(S!("OLS30402"))),
+                Some(EXTENSION_NAME.to_string()), 
+                format!("menuitem node must contain an id attribute"), 
+                None, 
+                None));
+        }
+        for child in node.children() {
+            if child.is_text() || child.is_comment() {
+                continue;
+            }
```

Comment:
```suggestion
        for child in node.children().filter(|n| n.is_element()) {
```

You do that later in the code anyway, it is shorter and consistent